Set browser language to initial state

marianhlavac opened the pull request:
Resolves #50

marianhlavac committed:
Use browser locale for initial settings

jnv commented:
I don't think this is the best way to handle this since you are introducing external effects to reducer. See 473fbad7f04a425305935cefd62fa11dff3bd22f – what do you think?

marianhlavac commented:
Oops, too long commit message, sry.

marianhlavac commented:
Ok, I've removed the side effects from the reducer and changed the browserLanguage function to accept directy the language string instead of window object.

jnv commented on src/reducers/index.js:
```diff
@@ -13,7 +13,7 @@ import client from './clientReducer'
 
 const initialSettings = {
-  locale: 'cs',
+  locale: null,
```
Don't do this. If you don't pass the initial state, the app will be broken.

marianhlavac replied:
!fixed

jnv commented on test/client.test.js:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  forEach(([lang, strings]) => {
+    console.log(lang, strings)
```
Remove console.log, please.

marianhlavac replied:
ooooops.

marianhlavac replied:
!fixed

jnv commented on src/Root.jsx:
```diff
@@ -4,8 +4,14 @@ import { Provider } from 'react-redux'
-const store = createStore()
+const initialState = {
+  settings: {
+    preferredLanguage: browserLanguage(window.navigator.userLanguage || window.navigator.language),
```
Why is this `preferredLanguage` and not `locale`?

marianhlavac replied:
Mistake, good you've noticed.

marianhlavac replied:
!resolved

jnv commented:
Have you tested what goes into initial settings without local storage? I am afraid you will get just locale property under settings reducer.

marianhlavac committed:
Move settings reducer into a separate state

marianhlavac commented:
Behaviour, when the localStorage is undefined, is fixed now.

jnv commented on src/reducers/settingsReducer.js:
```diff
@@ -0,0 +1,19 @@
+  facultyGrid: false,
+}
+
+export function reducer (state = initialState, action) {
```
Use `export default` to maintain modules compatibility.
!fix

marianhlavac replied:
!fixed

jnv commented on src/client.js:
```diff
@@ -0,0 +1,9 @@
+import { is } from 'ramda'
+
+export function browserLanguage (navigatorLanguage = 'cs') {
```
This gives out mixed signals: you want to fall back to Czech for undefined states, and fall back to English otherwise. This is okay, but you need to document it.
Add comment to the function, which explains this behaviour. !resolve

marianhlavac replied:
!resolved

jnv commented on test/client.test.js:
```diff
@@ -0,0 +1,18 @@
+
+  forEach(([lang, strings]) => {
+    forEach(string => {
+      t.equal(c.browserLanguage(string), lang, `returns ${lang} on browser lang ${string}`)
```
- Assign output of browserLanguage to `actual` const
- rename `lang` to `expectedLang`
- rename `string` to `testStr`
  !fix

marianhlavac replied:
!fixed

jnv commented:
I will squash the commits and merge this.
